### PR TITLE
[Problem Statement] TEP-0096: Pipelines V1 Release

### PR DESCRIPTION
**Pull request by lbernick**
This commit defines the problem statement for a stable release
of Tekton pipelines. It also describes the current Pipelines
deprecation policy and proposes a deprecation policy for a stable
release of Pipelines.

Co-authored-by: Jerop Kipruto jerop@google.com @jerop

### CODE REVIEW

File: teps/0096-pipelines-v1-release.md
Comment (vdemeester) [inline teps/0096-pipelines-v1-release.md:32]: By API definition, do we mean the "spec" of each object we ship ? I think that's could be intersting indeed 👼🏼 
Or are we only talking about "the policy" ?

File: teps/0096-pipelines-v1-release.md
Comment (vdemeester) [inline teps/0096-pipelines-v1-release.md:46]: Why is "Pipelines API definition" linking the api compatibility policy ? I would assume it would link the "spec reference" docs (the types or something).

File: teps/0096-pipelines-v1-release.md
Comment (vdemeester) [inline teps/0096-pipelines-v1-release.md:51]: What if some of those aspect are requirement we want to set before doing a V1 ? (dogfooding usage and performance are definitely two requirements I would personally make)

File: teps/0096-pipelines-v1-release.md
Comment (vdemeester) [inline teps/0096-pipelines-v1-release.md:64]: I am not sure I understand what is the relation between this (release *tags*) and api version. Nothing should prevent use to do alpha and rc releases, and I don't see any "relation" to the API version, stability, compatibility, …
Doing a release, in alpha, or rc, is essentially stating, their might be some bugs, and code that will change, but *most likely* the featureset won't change. It's unrelated to APIs, it's just a matter of what we ship, when and with what level of confidence, isn't it ? 👼🏼

File: teps/0096-pipelines-v1-release.md
Comment (vdemeester) [inline teps/0096-pipelines-v1-release.md:79]: When we say "Backwards incompatible API changes must be accompanied by a major version software release", do we means v2.x.y for example ? With APIs version, we should be able to provide a backward incompatible API (`v2alpha1` for example) independently of the major version in the release, right ? I would tend to agree though, that if we were to publish a `v2` API version, it should most likely require a major version.

**Review by vdemeester — COMMENTED** [review]
I am guessing this is the TEP version of https://github.com/tektoncd/pipeline/issues/3548 ?

/assign

File: teps/0096-pipelines-v1-release.md
Comment (lbernick) [reply teps/0096-pipelines-v1-release.md:46]: Updated to link to the "What is the API" heading

File: teps/0096-pipelines-v1-release.md
Comment (lbernick) [reply teps/0096-pipelines-v1-release.md:32]: I think what we consider the Pipelines API will be decided in a later update to this TEP, but our current answer to this question is [here](https://github.com/tektoncd/pipeline/blob/main/api_compatibility_policy.md#what-is-the-api). I am referring here to the objects we ship, but not specifically to their "spec" fields.

File: teps/0096-pipelines-v1-release.md
Comment (lbernick) [reply teps/0096-pipelines-v1-release.md:51]: What I meant here was that defining stability policies for these features is out of scope, not necessarily that such features can't be prioritized for V1. For example, I think it's fine to set some performance goals for V1, but out of scope to define performance benchmarks that we will meet for stable releases. I've updated the wording here to make this more clear.

File: teps/0096-pipelines-v1-release.md
Comment (lbernick) [reply teps/0096-pipelines-v1-release.md:79]: Added an example. In light of TEP 33, if we want to add an experimental feature, for example, we would add it behind an alpha feature flag, and it could be removed without changing the API version or software major version number. However, if we changed or removed a stable feature, that would require updating the API version from V1 to V2.

File: teps/0096-pipelines-v1-release.md
Comment (lbernick) [reply teps/0096-pipelines-v1-release.md:64]: I included information on software versioning here because I think it's easy to mix up software versioning with API versioning. This line was simply meant to be a description of the current state of the world, as I didn't see any "alpha", "beta", or "rc" versions when I looked through Tekton releases. This TEP is proposing to continue defining what is "breaking" or "backwards incompatible" based on the stability level of the API features, and to increment the software version accordingly based on semantic versioning. I am not sure I fully understand your question, but I hope this plus additional clarifying comments in the TEP answers it?

**Comment by lbernick** [comment]
> I am guessing this is the TEP version of [tektoncd/pipeline#3548](https://github.com/tektoncd/pipeline/issues/3548) ?

The goal of this TEP is to come to consensus on a stability policy and what features we'd like to prioritize for V1, but not to do detailed issue tracking here. Jerop and I are thinking that this will be similar to [TEP-0074: Deprecate PipelineResources](https://github.com/tektoncd/community/blob/main/teps/0074-deprecate-pipelineresources.md).

File: teps/0096-pipelines-v1-release.md
Comment (vdemeester) [reply teps/0096-pipelines-v1-release.md:64]: It does and it doesn't at the same time 😛 . 

> I included information on software versioning here because I think it's easy to mix up software versioning with API versioning

In the k8s ecosystem, I am not sure that's true. I might be wrong here, but it seems very clear that the kubernetes object API version is orthogonal to the release version. API `v1` for Knative was published before Knative did a `v1` release (tagged v1), similar with internal k8s object.

> Instead of doing an "alpha" release, we would keep the API version at V1, add an alpha feature behind an alpha feature flag, and do a software minor version release.

This doesn't prevent us from doing alpha and rc releases though. We could release an v1.1.0-alpha release that provides  some new field behind an alpha feature-gate, and move to v1.1.0-rc and v1.1.0 final while still keeping this new field behind an alpha feature flag. This is the same upstream with Kubernetes tbh. The "alpha"/"rc" state of a given release is about the code, not about the API version and status, as we version APIs..

Comment (lbernick) [reply teps/0096-pipelines-v1-release.md]: Thanks for the feedback! I've made software versioning a non goal of this TEP.

File: teps/0096-pipelines-v1-release.md
Comment (lbernick) [reply teps/0096-pipelines-v1-release.md]: moved to non goals!

File: teps/0096-pipelines-v1-release.md
Comment (afrittoli) [inline teps/0096-pipelines-v1-release.md:48]: Perhaps an extra one could be, define what updates, if any, are needed by the feature gates spec (the TEP) and tooling.

**Review by afrittoli — APPROVED** [review]
Thanks for starting this!

**Comment by lbernick** [comment]
@vdemeester @pritidesai @bobcatfish I'd love to be able to merge this before the next WG (12/13) because I'm hoping to start discussing in scope/out of scope features during that meeting! PTAL and let me know if you have any concerns or if you need more time with this proposal, but I'm hoping since there aren't really any changes proposed to what we're already doing, we can merge this soon 🤞

**Comment by xchapter7x** [comment]
@vdemeester / @iancoffey 
mind giving this one a quick review? :)

File: teps/0096-pipelines-v1-release.md
Comment (pritidesai) [inline teps/0096-pipelines-v1-release.md:32]: I will rephrase this as Tekton Pipelines provides Alpha (no more maintained or a better term here) and Beta APIs.

File: teps/0096-pipelines-v1-release.md
Comment (pritidesai) [reply teps/0096-pipelines-v1-release.md:34]: Yup, I read it as "spec" of CRDs as well. You might want to replace it to: This TEP defines Tekton Pipelines V1 API and its stability/compatibility policies ... its just a nit.

**Review by bobcatfish — COMMENTED** [review]
Looking good! just some minor comments from me

/approve

File: teps/0096-pipelines-v1-release.md
Comment (pritidesai) [inline teps/0096-pipelines-v1-release.md:41]: NIT: I will rephrase this to - Some users may be waiting for the V1 stability to adopt Tekton as their CI/CD platform in Production.

File: teps/0096-pipelines-v1-release.md
Comment (vdemeester) [reply teps/0096-pipelines-v1-release.md:46]: I think we may want to have both. Feature is a "limited" set that we know we have or want to have. Use case is.. something that is outside of our control (someone could be using it for a use case we never thought of) but we could definitely clarify the use case we intend to support (non exhaustive list 😛 ) and the one we definitely do not want to support (if any)

File: teps/0096-pipelines-v1-release.md
Comment (vdemeester) [reply teps/0096-pipelines-v1-release.md:55]: It doesn't match 100%, but close enough for me to agree with what you said 😉

**Review by vdemeester — APPROVED** [review]
Overall looks good, approving the problem statement. We have time to figure somethings out during the "design detail" phase 😉

File: teps/0096-pipelines-v1-release.md
Comment (pritidesai) [inline teps/0096-pipelines-v1-release.md:52]: +1 to this

but at the same time, can the other Tekton projects play any role in Pipelines V1? 🤔  we discussed briefly this in API WG, there could be but we don't have concrete answer yet or we don't know yet.

File: teps/0096-pipelines-v1-release.md
Comment (pritidesai) [inline teps/0096-pipelines-v1-release.md:54]: I am not clear what this alludes to. Apart from the release automation, most of these aspects can impact or be part of the stability policy, isn't it? For example, we might have to tweak the existing design if Tekton Pipelines is performing poor.

File: teps/0096-pipelines-v1-release.md
Comment (pritidesai) [inline teps/0096-pipelines-v1-release.md:67]: This might change once we release V1, can it? The focus is months right now since Pipelines release are created every month. What is our plan to create release once we hit V1 API and/or V1 software versioning? We don't have to answer this here but I am just wondering while reading this since its being documented officially 🤣  unless I am missing the existing documentation.

File: teps/0096-pipelines-v1-release.md
Comment (pritidesai) [inline teps/0096-pipelines-v1-release.md:3]: NIT: After reading the proposal, I think the emphasis is on Pipelines V1 API, "Release" might not happen months after V1 API is implemented.

**Comment by pritidesai** [comment]
Thank you @lbernick for creating a formal proposal for this and additional shout out to all the contributors who have played a key role in getting us so far. 

I have left some NITs and more discussion points but since this is an iterative proposal (process), I am happy to merge it 🎉 

/approve

File: teps/0096-pipelines-v1-release.md
Comment (lbernick) [reply teps/0096-pipelines-v1-release.md:3]: good point :)

File: teps/0096-pipelines-v1-release.md
Comment (lbernick) [reply teps/0096-pipelines-v1-release.md]: I think how often we want to create software releases is probably out of scope for the TEP, but we could choose to adopt Kubernetes language of "x months or x releases, whichever is longer". I don't see a strong reason in either direction; I think it's unlikely our release cadence would become significantly slower after a v1 release.

File: teps/0096-pipelines-v1-release.md
Comment (lbernick) [reply teps/0096-pipelines-v1-release.md]: The idea here was really to capture the non goals in the [original v1 doc](https://docs.google.com/document/d/1h9n0Lod0OiJ_sP2HK8Ms7N04aee5LW8xfz5yMGCFMIs/edit#), but it sounds like performance may be something we do want to consider explicitly for v1. @vdemeester I think you had a similar comment; thoughts on moving performance out of non goals? Priti, would you want this non goal removed entirely or just reworded?

File: teps/0096-pipelines-v1-release.md
Comment (lbernick) [reply teps/0096-pipelines-v1-release.md]: I think other Tekton projects will be important for meeting the use cases we want to support, but we can define what it means for pipelines v1 to be stable without involving features from other projects (hopefully). This seems like something that should remain an open question for now.

File: teps/0096-pipelines-v1-release.md
Comment (lbernick) [reply teps/0096-pipelines-v1-release.md]: Priti had a similar comment below, so I'm just going to close this one so the discussion stays in one place!

File: teps/0096-pipelines-v1-release.md
Comment (pritidesai) [reply teps/0096-pipelines-v1-release.md]: I think the original V1 doc described these as non-goals to not restrict V1 to include these aspects. They could be very well related to the stability policies of the API. I think we can drop it for now as a non-goal and capture it in the next iterations. The time we captured these non-goals in the original V1 doc v/s now, there has been many enhancements and updates done, we might have to revisit based on the latest status of the project.

Comment (lbernick) [reply teps/0096-pipelines-v1-release.md]: sounds good, I've removed this as a non goal.

**Comment by pritidesai** [comment]
Thank you @lbernick for addressing all the comments, we have received approvals from all the assignees. Its ready to merge 🎉 

/lgtm

**Pull request merged**
